File: kotlin/src/test/kotlin/aethernet/circuitrelay/RelayEngineTest.kt
// SPDX-License-Identifier: MIT

package aethernet.circuitrelay

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.BlockingQueue
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Behavioural proof of the native circuit-relay-v2 engine ([Transport]): a three-node
 * topology where A and B can each reach relay R but NOT each other directly. A message
 * from A must traverse the relay bridge to reach B — server off, no libp2p. Mirrors the
 * Go `transport_test.go` (6 cases) and the C# `CircuitRelayBridgeTests`.
 *
 * No kotlinx.coroutines: the in-process link delivers frames on a cached thread pool
 * (like a real async transport, and to avoid re-entrant deadlock), and the test awaits
 * receipts / status via [BlockingQueue].
 */
class RelayEngineTest {

    // ── in-process one-hop mesh ──────────────────────────────────────────────

    private class Mesh(private val pool: ExecutorService) {
        private val lock = ReentrantLock()
        private val edges = HashSet<String>()
        private val links = HashMap<String, ProcLink>()

        fun connect(x: String, y: String) = lock.withLock {
            edges.add("$x|$y")
            edges.add("$y|$x")
        }

        fun adjacent(x: String, y: String): Boolean = lock.withLock { edges.contains("$x|$y") }

        fun link(node: String): ProcLink = lock.withLock {
            links.getOrPut(node) { ProcLink(this, node) }
        }

        fun deliver(from: String, to: String, frame: ByteArray) {
            if (!adjacent(from, to)) return
            val l = link(to)
            pool.execute { // async hop, like a real transport
                l.handler()?.invoke(from, frame)
            }
        }
    }

    private class ProcLink(private val mesh: Mesh, private val node: String) : RelayLink {
        private val hlock = ReentrantLock()
        private var h: ((String, ByteArray) -> Unit)? = null

        fun handler(): ((String, ByteArray) -> Unit)? = hlock.withLock { h }

        override fun sendFrame(node: String, frame: ByteArray): Boolean {
            if (!mesh.adjacent(this.node, node)) return false
            mesh.deliver(this.node, node, frame)
            return true
        }

        override fun canReach(node: String): Boolean = mesh.adjacent(this.node, node)

        override fun onFrame(handler: (String, ByteArray) -> Unit) {
            hlock.withLock { h = handler }
        }
    }

    // ── controllable clock ───────────────────────────────────────────────────

    private class TestClock {
        private val lock = ReentrantLock()

        // Start at 2026-01-01T00:00:00Z, in epoch ms.
        private var t: Long = 1_767_225_600_000L

        fun now(): Long = lock.withLock { t }
        fun advanceMs(d: Long) = lock.withLock { t += d }
    }

    private data class Recv(val sender: String, val data: String)

    // ── shared fixture: A ── R ── B with NO A-B edge ─────────────────────────

    private lateinit var pool: ExecutorService

    private class Line(
        val a: Transport,
        val r: Transport,
        val b: Transport,
        val bRecv: BlockingQueue<Recv>,
        val aRecv: BlockingQueue<Recv>
    )

    /** Wires A ── R ── B with no A-B edge. [relayOpts]/[relayClock] configure R. */
    private fun buildLine(
        relayOpts: RelayOptions = RelayOptions(),
        relayClock: (() -> Long)? = null
    ): Line {
        val mesh = Mesh(pool)
        mesh.connect("A", "R")
        mesh.connect("R", "B")
        val a = Transport("A", mesh.link("A"))
        val r = if (relayClock != null) {
            Transport("R", mesh.link("R"), relayOpts, relayClock)
        } else {
            Transport("R", mesh.link("R"), relayOpts)
        }
        val b = Transport("B", mesh.link("B"))
        val bRecv: BlockingQueue<Recv> = ArrayBlockingQueue(8)
        val aRecv: BlockingQueue<Recv> = ArrayBlockingQueue(8)
        b.setOnData { s, d -> bRecv.offer(Recv(s, String(d, Charsets.UTF_8))) }
        a.setOnData { s, d -> aRecv.offer(Recv(s, String(d, Charsets.UTF_8))) }
        return Line(a, r, b, bRecv, aRecv)
    }

    private fun waitRecv(ch: BlockingQueue<Recv>, what: String): Recv =
        ch.poll(3, TimeUnit.SECONDS) ?: error("timeout waiting for $what")

    @org.junit.jupiter.api.BeforeEach
    fun setUp() {
        pool = Executors.newCachedThreadPool()
    }

    @AfterEach
    fun tearDown() {
        pool.shutdownNow()
        pool.awaitTermination(2, TimeUnit.SECONDS)
    }

    // ── (a) A -> R -> B relay, B receives, relay bridge count == 1 ───────────

    @Test
    fun `message traverses relay with no direct link`() {
        val line = buildLine()

        assertFalse(line.a.isConnected("B"), "A should not be directly connected to B")
        assertTrue(line.b.reserve("R"), "B.reserve(R) failed")
        line.a.setRoute("B", "R")

        assertTrue(line.a.send("B", "deadbeef".toByteArray(Charsets.UTF_8)), "A.send returned false")

        val got = waitRecv(line.bRecv, "B receiving relayed message")
        assertEquals("A", got.sender, "sender")
        assertEquals("deadbeef", got.data, "payload")
        assertEquals(1, line.r.activeBridgeCount(), "relay bridge count")
    }

    // ── (b) bidirectional ────────────────────────────────────────────────────

    @Test
    fun `bridge is bidirectional`() {
        val line = buildLine()
        assertTrue(line.b.reserve("R"), "reserve failed")
        line.a.setRoute("B", "R")
        assertTrue(line.a.send("B", "hi".toByteArray(Charsets.UTF_8)), "A.send failed")
        waitRecv(line.bRecv, "B receiving")

        assertTrue(line.b.send("A", "reply".toByteArray(Charsets.UTF_8)), "B.send(A) failed")
        val got = waitRecv(line.aRecv, "A receiving B's reply")
        assertEquals("B", got.sender, "sender")
        assertEquals("reply", got.data, "payload")
    }

    // ── (c) connect refused without reservation ──────────────────────────────

    @Test
    fun `connect refused without reservation`() {
        val line = buildLine()
        line.a.setRoute("B", "R") // route known, but B never reserved

        assertFalse(line.a.send("B", "x".toByteArray(Charsets.UTF_8)), "A.send should fail without a reservation")

        assertNull(line.bRecv.poll(200, TimeUnit.MILLISECONDS), "B should not have received anything")
        assertEquals(0, line.r.activeBridgeCount(), "relay bridge count")
    }

    // ── (d) send fails with no route ─────────────────────────────────────────

    @Test
    fun `send fails without route`() {
        val line = buildLine()
        assertTrue(line.b.reserve("R"), "reserve failed")
        // no setRoute
        assertFalse(line.a.send("B", "x".toByteArray(Charsets.UTF_8)), "A.send should fail with no relay route known")
    }

    // ── (e) data budget 10: first 5B delivered, second 8B (cum 13) dropped ───

    @Test
    fun `relay enforces data budget`() {
        val opts = RelayOptions(bridgeDataLimitBytes = 10)
        val line = buildLine(opts)
        assertTrue(line.b.reserve("R"), "reserve failed")
        line.a.setRoute("B", "R")

        assertTrue(
            line.a.send("B", byteArrayOf(1, 2, 3, 4, 5)), // 5 bytes, within 10
            "first send failed"
        )
        waitRecv(line.bRecv, "first (in-budget) message")

        line.a.send("B", byteArrayOf(6, 7, 8, 9, 10, 11, 12, 13)) // 8 more -> 13 > 10 -> torn down
        assertNull(line.bRecv.poll(300, TimeUnit.MILLISECONDS), "over-budget message should not arrive")
        assertEquals(0, line.r.activeBridgeCount(), "bridge should be torn down on budget breach")
    }

    // ── (f) reservation expiry via injectable clock ──────────────────────────

    @Test
    fun `reservation expiry refuses connect`() {
        val clk = TestClock()
        val opts = RelayOptions(reservationTtlMs = 30L * 60L * 1000L)
        val line = buildLine(opts, clk::now)

        assertTrue(line.b.reserve("R"), "reserve failed")
        line.a.setRoute("B", "R")

        clk.advanceMs(31L * 60L * 1000L) // past the reservation TTL on R's clock

        assertFalse(line.a.send("B", "x".toByteArray(Charsets.UTF_8)), "A.send should fail after reservation expiry")
        assertNull(line.bRecv.poll(200, TimeUnit.MILLISECONDS), "B should not receive after expiry")
    }
}

File: typescript/tests/relay_engine.test.ts
// SPDX-License-Identifier: MIT
//
// Behavioural proof of the native circuit-relay-v2 ENGINE: a three-node topology
// where A and B can each reach relay R but NOT each other directly. A message from
// A must traverse the relay bridge to reach B — server off, no libp2p. Mirrors the
// Go transport_test.go (in-process 3-node mesh + 6 behavioural cases) and the C#
// CircuitRelayBridgeTests.

import test from "node:test";
import assert from "node:assert/strict";

import {
  Transport,
  RelayLink,
  RelayOptions,
  defaultRelayOptions,
} from "../src/circuitrelay/Transport.js";

// ── in-process one-hop mesh ──────────────────────────────────────────────────

class Mesh {
  private readonly edges = new Set<string>();
  private readonly links = new Map<string, ProcLink>();

  connect(x: string, y: string): void {
    this.edges.add(x + "|" + y);
    this.edges.add(y + "|" + x);
  }

  adjacent(x: string, y: string): boolean {
    return this.edges.has(x + "|" + y);
  }

  link(node: string): ProcLink {
    let l = this.links.get(node);
    if (l === undefined) {
      l = new ProcLink(this, node);
      this.links.set(node, l);
    }
    return l;
  }

  deliver(from: string, to: string, frame: Uint8Array): void {
    if (!this.adjacent(from, to)) return;
    const l = this.link(to);
    // async hop, like a real transport — avoids re-entrant recursion.
    setTimeout(() => l.dispatch(from, frame), 0);
  }
}

class ProcLink implements RelayLink {
  private handler: ((from: string, frame: Uint8Array) => void) | null = null;

  constructor(
    private readonly mesh: Mesh,
    private readonly node: string,
  ) {}

  sendFrame(node: string, frame: Uint8Array): boolean {
    if (!this.mesh.adjacent(this.node, node)) return false;
    this.mesh.deliver(this.node, node, frame);
    return true;
  }

  canReach(node: string): boolean {
    return this.mesh.adjacent(this.node, node);
  }

  onFrame(handler: (from: string, frame: Uint8Array) => void): void {
    this.handler = handler;
  }

  dispatch(from: string, frame: Uint8Array): void {
    this.handler?.(from, frame);
  }
}

// ── controllable clock ───────────────────────────────────────────────────────

class TestClock {
  // 2026-01-01T00:00:00Z, in epoch ms.
  private t = Date.UTC(2026, 0, 1, 0, 0, 0, 0);
  now = (): number => this.t;
  advance(ms: number): void {
    this.t += ms;
  }
}

// ── receipt helper: resolve with the next delivered message, or reject on timeout ─

interface Recv {
  sender: string;
  data: string;
}

/** A per-transport receipt queue with a promise+timeout awaiter, mirroring the Go `chan recv`. */
class RecvQueue {
  private readonly buffer: Recv[] = [];
  private waiter: ((r: Recv) => void) | null = null;

  push = (sender: string, data: Uint8Array): void => {
    const r: Recv = { sender, data: new TextDecoder().decode(data) };
    if (this.waiter) {
      const w = this.waiter;
      this.waiter = null;
      w(r);
    } else {
      this.buffer.push(r);
    }
  };

  /** Await the next message; reject after `timeoutMs`. */
  next(what: string, timeoutMs = 3000): Promise<Recv> {
    const queued = this.buffer.shift();
    if (queued !== undefined) return Promise.resolve(queued);
    return new Promise<Recv>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.waiter = null;
        reject(new Error(`timeout waiting for ${what}`));
      }, timeoutMs);
      this.waiter = (r) => {
        clearTimeout(timer);
        resolve(r);
      };
    });
  }

  /** Assert NO message arrives within `windowMs`. */
  async expectNone(windowMs: number): Promise<void> {
    await sleep(windowMs);
    const got = this.buffer.shift();
    if (got !== undefined) {
      throw new Error(`unexpected message: {${got.sender} ${got.data}}`);
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

interface Line {
  a: Transport;
  r: Transport;
  b: Transport;
  bRecv: RecvQueue;
  aRecv: RecvQueue;
}

// buildLine wires A ── R ── B with NO A-B edge. relayOpts/relayClock configure R.
function buildLine(relayOpts: RelayOptions, relayClock?: () => number): Line {
  const m = new Mesh();
  m.connect("A", "R");
  m.connect("R", "B");
  const a = new Transport("A", m.link("A"), defaultRelayOptions());
  const r = new Transport("R", m.link("R"), relayOpts, relayClock);
  const b = new Transport("B", m.link("B"), defaultRelayOptions());
  const bRecv = new RecvQueue();
  const aRecv = new RecvQueue();
  b.setOnData(bRecv.push);
  a.setOnData(aRecv.push);
  return { a, r, b, bRecv, aRecv };
}

// (a) A->R->B relay, B receives; R activeBridgeCount == 1.
test("engine: message traverses relay with no direct link", async () => {
  const { a, r, b, bRecv } = buildLine(defaultRelayOptions());

  assert.equal(a.isConnected("B"), false, "A should not be directly connected to B");
  assert.equal(await b.reserve("R"), true, "B.reserve(R) failed");
  a.setRoute("B", "R");

  assert.equal(await a.send("B", new TextEncoder().encode("deadbeef")), true, "A.send returned false");

  const got = await bRecv.next("B receiving relayed message");
  assert.equal(got.sender, "A");
  assert.equal(got.data, "deadbeef");
  assert.equal(r.activeBridgeCount(), 1, "relay bridge count should be 1");
});

// (b) bidirectional.
test("engine: bridge is bidirectional", async () => {
  const { a, b, bRecv, aRecv } = buildLine(defaultRelayOptions());
  assert.equal(await b.reserve("R"), true, "reserve failed");
  a.setRoute("B", "R");
  assert.equal(await a.send("B", new TextEncoder().encode("hi")), true, "A.send failed");
  await bRecv.next("B receiving");

  assert.equal(await b.send("A", new TextEncoder().encode("reply")), true, "B.send(A) failed");
  const got = await aRecv.next("A receiving B's reply");
  assert.equal(got.sender, "B");
  assert.equal(got.data, "reply");
});

// (c) connect refused without reservation.
test("engine: connect refused without reservation", async () => {
  const { a, r, bRecv } = buildLine(defaultRelayOptions());
  a.setRoute("B", "R"); // route known, but B never reserved
  assert.equal(await a.send("B", new TextEncoder().encode("x")), false, "A.send should fail without a reservation");
  await bRecv.expectNone(200);
  assert.equal(r.activeBridgeCount(), 0, "relay bridge count should be 0");
});

// (d) send fails with no route.
test("engine: send fails without route", async () => {
  const { a, b } = buildLine(defaultRelayOptions());
  assert.equal(await b.reserve("R"), true, "reserve failed");
  // no setRoute
  assert.equal(await a.send("B", new TextEncoder().encode("x")), false, "A.send should fail with no relay route known");
});

// (e) data budget 10 -> first 5B delivered, second 8B (cum 13) dropped + bridge torn down.
test("engine: relay enforces data budget", async () => {
  const opts = defaultRelayOptions();
  opts.bridgeDataLimitBytes = 10;
  const { a, r, b, bRecv } = buildLine(opts);
  assert.equal(await b.reserve("R"), true, "reserve failed");
  a.setRoute("B", "R");

  assert.equal(await a.send("B", new Uint8Array([1, 2, 3, 4, 5])), true, "first send failed"); // 5 bytes, within 10
  await bRecv.next("first (in-budget) message");

  await a.send("B", new Uint8Array([6, 7, 8, 9, 10, 11, 12, 13])); // 8 more -> 13 > 10 -> torn down
  await bRecv.expectNone(300);
  assert.equal(r.activeBridgeCount(), 0, "bridge should be torn down on budget breach");
});

// (f) reservation expiry via injectable clock.
test("engine: reservation expiry refuses connect", async () => {
  const clk = new TestClock();
  const opts = defaultRelayOptions();
  opts.reservationTtlMs = 30 * 60 * 1000; // 30 minutes
  const { a, b, bRecv } = buildLine(opts, clk.now);

  assert.equal(await b.reserve("R"), true, "reserve failed");
  a.setRoute("B", "R");

  clk.advance(31 * 60 * 1000); // past the reservation TTL on R's clock

  assert.equal(await a.send("B", new TextEncoder().encode("x")), false, "A.send should fail after reservation expiry");
  await bRecv.expectNone(200);
});
